### ISSUE TITLE
Optimal memory allocation with DPDK for nDPI

### ISSUE THREAD

**Issue by DimaWittmann** [ntop/nDPI#1781]
Dear nDPI team,
Thank you for your work and time.

I was doing a performance test of our application, it's based on DPDK with nDPI integrated.
I noticed that malloc is a bottleneck, essentially it's expected. So I decided to use memory pools to replace rte_malloc for nDPI.
My question is about optimal object sizes for the pools. I tried to have 4 memory pools: 64B, 512B, 4KB, and 32KB and it worked out pretty well. I wondered if you had a similar experience and have some information about optimal sizes.
Another question is how many objects are allocated per traffic flow, to estimate the needed size of the pools.
Please share your thoughts and assumption.

Thank you very much.

**Comment by utoni** [comment]
This is interesting. I was also thinking about mempools for `nDPId`s implementation. But I was too scarred about a possible memory waste. Can you tell something about the usage of the mempool?
~~How do you handle memory fragmentation?~~ Edit: I presume that your mempool chunks have a fixed size, so fragmentation is not an issue.

I guess it is useful to have a pool size where the ndpi flow struct fits perfectly in and some smaller sizes e.g. for string allocation.

**Comment by IvanNardi** [comment]
> I noticed that malloc is a bottleneck, essentially it's expected.

I am very surprised by that.
In my experience, once you pre-allocate `struct ndpi_flow_struct` structures, memory allocation is no more an issue.
I mean, there is still a lot of room for improvements, but memory is definitely not the main performance bottleneck, to me.

Could you share some info on your environment? Traffic rate, hardware... Are you using plain `ndpiReader` with DPDK support?

**Comment by DimaWittmann** [comment]
The setup is following:
15 cores are dedicated to packet processing, essentially they do some basic packet parsing, and then a call to _ndpi_detection_process_packet_
Four 10G ports.
The flow table is about 8+ million entries.

Initially, I did not provide malloc to nDPI, so libc malloc was used, and I did not notice that malloc was the bottleneck. The problem was that nDPI allocated much memory besides ` struct ndpi_flow_struct,` about 1GB per 1 million entries. The lion part of the memory in our platform is allocated for DPDK as huge pages, so I provided _rte_malloc_ to nDPI so that huge pages will be used. 
With more than 10G of traffic, I see significant performance degradation, particularly with `rte_malloc`. Profiling data shows that rte_malloc-related functions take about 60%-80% of the processing time under load.

Regarding the memory pools, the picture is similar to what Toni suggests, a memory pool for `struct ndpi_flow_struct` and four memory pools: 64B, 512B, 4KB, and 32KB for strings or whatever.
Maybe you know how string sizes depend on the packet size or on the protocol? Is there some correlation?

**Comment by IvanNardi** [comment]
> about 1GB per 1 million entries.

I don't have any numbers, but I am asking a very silly question (sorry for that): are you aware that you can free `struct ndpi_flow_struct` (and all related memory) once the flow is classified (i.e. usually after 1-5 pkts)? In other words, you should usually have:
numbers of `struct ndpi_flow_struct` structures "used" << 8 million

This way the total memory used should be lower; no idea of the impact on the performance, though.

Just checking the basic...

> 64B, 512B, 4KB, and 32KB

While 64 and 512 might have sense for the string or basic buffer, and 4K might be used for reassembler buffer, I am a little surprised that you need chunks of memory bigger that 4K... is the application really using them?

Another basic question: is the traffic the "standard" Internet/enterprise traffic? DNS/TLS/QUIC/...

**Comment by DimaWittmann** [comment]
Thank you, this is a good hint, I did not clean flows before flow expiration or related data. It can definitely improve the situation.
I see allocations about 30-40KB depending on traffic but I decided to abolish 512B and 32KB mempools, 512B does not look like it's much used and 32KB just waste too much memory.

For testing I used examples from nDPI, like tls_*, facebook*, whatsup*, youtube*, telegram pcaps

**Comment by IvanNardi** [comment]
With the simple, but ugly, patch attached you can easily look at the size of the memory allocations by the library in the "data path"
[mem_size.patch.txt](https://github.com/ntop/nDPI/files/9854462/mem_size.patch.txt)

There are some bigger allocations that expected.

```
ivan@ivan-Latitude-E6540:~/svnrepos/nDPI(dev)$ ./tests/do.sh | grep Memorysize | tr -d \Memorysize: | sort -n
[...]
 6468
 6468
 6468
 6468
 6477
 6676
 6752
 7261
 10165
 12996
```
It seems that most (all?) of this big allocations are related to (TLS) reassembler; definitely an area to investigate in the future.

Anyway, from that, you can easily create an histogram of allocation size, to properly allocate the pools.
Keep in mind that, while the files under `tests\pcaps` are extremely useful as ..tests, they might not be a good sample of generic traffic

I hope it will be of some use

**Comment by DimaWittmann** [comment]
Thank you very much, 
for now, it looks like the big allocations occur only once per some number of flows. Definitely, it's traffic related but with big packets, less of contention is expected, so it does not look like it is worth using an additional memory pool for big chunks.

**Comment by IvanNardi** [comment]
Basic malloc profiling added in [db9f6ec](https://github.com/ntop/nDPI/commit/db9f6ec1b4018164e5bff05f115dc60711bb711b)

Closing it for now.
If you have other issues/questions, please open a new ticket.
